### PR TITLE
Refactor build stage

### LINKED ISSUE #623: Seperate optimisation logic from the rest of the build logic
## Description of issue / requirement to address

Currently the optimisation process for some BLUEPRINT components is mixed in with other steps needed to build a complete reactor component. This is a problem if one wishes to create a reactor component without optimisation, e.g. when recreating the results of a previous run by reading it in.

## Proposed solution

Separate the optimisation logic into callbacks stored on the reactor class. A user can then configure them to do different jobs, or none at all, as needed. By setting a callback to `None` optimisation can be turned off while the rest of the build should work as expected.

## Alternative solutions

Boolean logic and if-statements could be used for this but it would result in more complicated control-flow and be less flexible. Inheritance and overriding of methods could be used instead of callbacks but has more overhead in terms of how much code is required.

## Additional context

The STEP reactor project wants to be able to recreate a complete Reactor object in memory by reading in the output from a previous run. This will allow users to write scripts that produce any additional output they want, without having to modify the output methods in the main project.

### PR DESCRIPTION
**Pull request by cmacmackin**
## Linked Issues

Closes #623

Incorporates #563 

## Description

This PR attempts to put optimisation tasks for the BLUEPRINT system components into callbacks. This separates them from the rest of the build task and allows them to be overridden dynamically to change the optimisation or turn it off.

EDIT: It also makes some improvements to `compare_dicts` so that a tolerance can be specified by the caller and element-wise comparison works on dictionaries containing lists.

## Interface Changes

Each callback has its own interface, which can be found in the default implementations in `BLUEPRINT/systems/optimisation_callbacks.py`.

EDIT: I have added additional parameters to the `compare_dicts` method. The user can now specify what tolerances to use when `almost_equal` is True. The defaults will give the same behaviour as before.

## Checklist

I confirm that I have completed the following checks:

- [x] Tests run locally and pass `pytest tests --reactor`
- [x] Code quality checks run locally and pass `flake8` and `black .`
- [x] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

**Review by je-cook — CHANGES_REQUESTED** [review]
A few minor changes on this side

File: BLUEPRINT/systems/optimisation_callbacks.py
Comment (cmacmackin) [reply BLUEPRINT/systems/optimisation_callbacks.py:128]: Just to clarify, you want me to change the other use so it is also called `initial_profile`?

File: BLUEPRINT/systems/optimisation_callbacks.py
Comment (cmacmackin) [reply BLUEPRINT/systems/optimisation_callbacks.py:124]: Oops, think that came from me copying those definitions from another routine.

File: BLUEPRINT/systems/optimisation_callbacks.py
Comment (cmacmackin) [reply BLUEPRINT/systems/optimisation_callbacks.py:128]: If that is the case, should I also change the name of `make_preliminary_profile()`?

Comment (je-cook) [reply BLUEPRINT/systems/optimisation_callbacks.py:128]: Ok good point `FW.initial_profile` -> `FW.preliminary_profile` otherwise we will be changing the interface too much probably

Comment (cmacmackin) [reply BLUEPRINT/systems/optimisation_callbacks.py:128]: Done.

File: BLUEPRINT/systems/firstwall.py
Comment (cmacmackin) [reply BLUEPRINT/systems/firstwall.py:1161]: Should we delay merging until there is a comment on this?

Comment (DarioV86) [reply BLUEPRINT/systems/firstwall.py:1161]: Sorry, I read only now.
There is no explanation. I gave `hf_limit = 0.2` just to make the optimiser do something, otherwise with higher values it wouldn't change the wall profile.
Similarly `n_iteration_max=5`. Just a random value to try if the loop stops after 5 iterations.

Comment (je-cook) [reply BLUEPRINT/systems/firstwall.py:1161]: Ok, @DarioV86 can you make a PR once this PR is in to add a note to the function description (and at the same time move these to default args of the FW_optimisation function).

I'll let @sebkahn have a look at this but I'm happy to merge as is

**Review by je-cook — APPROVED** [review]

File: BLUEPRINT/reactor.py
Comment (sebkahn) [inline BLUEPRINT/reactor.py:719]:
```suggestion
        # NOTE: Should we just be relying on this mode instead?
```

File: bluemira/equilibria/profiles.py
Comment (sebkahn) [inline bluemira/equilibria/profiles.py:649]: ref @alexanderianblair for the scaling

File: bluemira/utilities/tools.py
Comment (sebkahn) [inline bluemira/utilities/tools.py:446]: Many thanks for that

**Review by sebkahn — CHANGES_REQUESTED** [review]

File: BLUEPRINT/systems/optimisation_callbacks.py
Comment (CoronelBuendia) [reply BLUEPRINT/systems/optimisation_callbacks.py:84]: not yet

File: bluemira/equilibria/equilibrium.py
Comment (CoronelBuendia) [reply bluemira/equilibria/equilibrium.py:707]: Some of these changes come in in #563 which has yet to be merged...

File: bluemira/equilibria/profiles.py
Comment (CoronelBuendia) [reply bluemira/equilibria/profiles.py:649]: #563 . We should merge that in first otherwise there will be conflicts and this PR is not related to this issue

File: BLUEPRINT/reactor.py
Comment (cmacmackin) [reply BLUEPRINT/reactor.py:719]: I think it's actually better to delete this entirely. It was a note to myself while working which won't be useful to anyone else.

File: BLUEPRINT/systems/tfcoils.py
Comment (cmacmackin) [reply BLUEPRINT/systems/tfcoils.py:704]: Removed the comment.

File: bluemira/equilibria/profiles.py
Comment (cmacmackin) [reply bluemira/equilibria/profiles.py:649]: Please merge that as soon as you can so we can get this merge completed, along with the one on the STEP repo. The longer we wait the more likely the latter-most will need to have more conflicts resolved.

**Review by sebkahn — APPROVED** [review]

**Pull request merged**
